src/checkbox_bubble.js

```diff
@@ -0,0 +1,227 @@
+   * @const
+   */
+  static CHECKMARK_PATH =
+    "M" +
+    CheckboxBubble.CHECKBOX_SIZE / 4 +
+    " " +
+    CheckboxBubble.CHECKBOX_SIZE / 2 +
+    "L" +
+    (5 * CheckboxBubble.CHECKBOX_SIZE) / 12 +
+    " " +
+    (2 * CheckboxBubble.CHECKBOX_SIZE) / 3 +
+    "L" +
+    (3 * CheckboxBubble.CHECKBOX_SIZE) / 4 +
+    " " +
+    CheckboxBubble.CHECKBOX_SIZE / 3;
+
+  /**
+   * Size of the checkbox corner radius
+   * @type {number}
+   * @const
+   */
+  static CHECKBOX_CORNER_RADIUS = 5;
+
+  /**
+   * @type {number}
+   * @const
+   */
+  static CHECKBOX_MARGIN = 12;
+
+  /**
+   * Total additional width of a row that contains a checkbox.
+   * @type {number}
+   * @const
+   */
+  static CHECKBOX_SPACE_X =
+    CheckboxBubble.CHECKBOX_SIZE + 2 * CheckboxBubble.CHECKBOX_MARGIN;
+
+  svgRoot;
+  clickListener;
+  checked = false;
+  location = new Blockly.utils.Coordinate(0, 0);
+
+  constructor(sourceBlock) {
+    this.sourceBlock = sourceBlock;
+    this.svgRoot = Blockly.utils.dom.createSvgElement(
+      Blockly.utils.Svg.G,
+      {},
+      this.sourceBlock.workspace.getBubbleCanvas()
+    );
+
+    const touchMargin = CheckboxBubble.CHECKBOX_TOUCH_PADDING;
+    const checkboxGroup = Blockly.utils.dom.createSvgElement(
+      "g",
+      {
+        fill: "transparent",
+      },
+      null
+    );
+    Blockly.utils.dom.createSvgElement(
+      "rect",
+      {
+        class: "blocklyFlyoutCheckbox",
+        height: CheckboxBubble.CHECKBOX_SIZE,
+        width: CheckboxBubble.CHECKBOX_SIZE,
+        rx: CheckboxBubble.CHECKBOX_CORNER_RADIUS,
+        ry: CheckboxBubble.CHECKBOX_CORNER_RADIUS,
+      },
+      checkboxGroup
+    );
+    Blockly.utils.dom.createSvgElement(
+      "path",
+      {
+        class: "blocklyFlyoutCheckboxPath",
+        d: CheckboxBubble.CHECKMARK_PATH,
+      },
+      checkboxGroup
+    );
+    Blockly.utils.dom.createSvgElement(
+      "rect",
+      {
+        class: "blocklyTouchTargetBackground",
+        x: -touchMargin + "px",
+        y: -touchMargin + "px",
+        height: CheckboxBubble.CHECKBOX_SIZE + 2 * touchMargin,
+        width: CheckboxBubble.CHECKBOX_SIZE + 2 * touchMargin,
+      },
+      checkboxGroup
+    );
+    this.setChecked(this.isChecked(this.sourceBlock.id));
+
+    this.svgRoot.prepend(checkboxGroup);
+
+    this.clickListener = Blockly.browserEvents.bind(
+      this.svgRoot,
+      "mousedown",
+      null,
+      (event) => {
+        this.setChecked(!this.checked);
+        event.stopPropagation();
+        event.preventDefault();
+      }
+    );
+    this.updateLocation();
+  }
+
+  setChecked(checked) {
+    this.checked = checked;
+    if (this.checked) {
+      Blockly.utils.dom.addClass(this.svgRoot, "checked");
+    } else {
+      Blockly.utils.dom.removeClass(this.svgRoot, "checked");
+    }
+
+    Blockly.Events.fire(
+      new Blockly.Events.BlockChange(
+        this.sourceBlock,
+        "checkbox",
+        null,
+        !this.checked,
```

Comment:
if these two params are the old state and the new state, you should check whether the checked state actually changed.

Reply:
Done